// agm-direction/index.js
import { Directive, EventEmitter, Input, NgModule, Output } from '@angular/core';
import { GoogleMapsAPIWrapper } from '@agm/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AgmDirection = (function () {
    function AgmDirection(gmapsApi) {
        this.gmapsApi = gmapsApi;
        this.travelMode = 'DRIVING';
        this.transitOptions = undefined;
        this.drivingOptions = undefined;
        this.waypoints = [];
        this.optimizeWaypoints = true;
        this.provideRouteAlternatives = false;
        this.avoidHighways = false;
        this.avoidTolls = false;
        this.visible = true;
        this.onChange = new EventEmitter();
        this.directionsService = undefined;
        this.directionsDisplay = undefined;
        this.isFirstChange = true;
        this.originMarker = undefined;
        this.destinationMarker = undefined;
    }
    /**
     * @return {?}
     */
    AgmDirection.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.directionDraw();
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    AgmDirection.prototype.ngOnChanges = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /**
             * When visible is false then remove the direction layer
             */
        if (!this.visible) {
            try {
                if (this.originMarker !== 'undefined') {
                    this.originMarker.setMap(null);
                    this.destinationMarker.setMap(null);
                }
                this.directionsDisplay.setPanel(null);
                this.directionsDisplay.setMap(null);
                this.directionsDisplay = undefined;
            }
            catch (/** @type {?} */ e) { }
        }
        else {
            if (this.isFirstChange) {
                this.isFirstChange = false;
                return;
            }
            /**
                   * When renderOptions are not first change then reset the display
                   */
            if (obj.renderOptions) {
                if (obj.renderOptions.firstChange === false) {
                    if (this.originMarker !== 'undefined') {
                        this.originMarker.setMap(null);
                        this.destinationMarker.setMap(null);
                    }
                    this.directionsDisplay.setPanel(null);
                    this.directionsDisplay.setMap(null);
                    this.directionsDisplay = undefined;
                }
            }
            this.directionDraw();
        }
    };
    /**
     * This event is fired when the user creating or updating this direction
     * @return {?}
     */
    AgmDirection.prototype.directionDraw = /**
     * This event is fired when the user creating or updating this direction
     * @return {?}
     */
    function () {
        var _this = this;
        this.gmapsApi.getNativeMap().then(function (map) {
            if (typeof _this.directionsDisplay === 'undefined') {
                _this.directionsDisplay = new google.maps.DirectionsRenderer(_this.renderOptions);
                _this.directionsDisplay.setMap(map);
            }
            if (typeof _this.directionsService === 'undefined') {
                _this.directionsService = new google.maps.DirectionsService;
            }
            if (typeof _this.panel === 'undefined') {
                _this.directionsDisplay.setPanel(null);
            }
            else {
                _this.directionsDisplay.setPanel(_this.panel);
            }
            _this.directionsService.route({
                origin: _this.origin,
                destination: _this.destination,
                travelMode: _this.travelMode,
                transitOptions: _this.transitOptions,
                drivingOptions: _this.drivingOptions,
                waypoints: _this.waypoints,
                optimizeWaypoints: _this.optimizeWaypoints,
                provideRouteAlternatives: _this.provideRouteAlternatives,
                avoidHighways: _this.avoidHighways,
                avoidTolls: _this.avoidTolls,
            }, function (response, status) {
                if (status === 'OK') {
                    _this.directionsDisplay.setDirections(response);
                    /**
                               * Emit The DirectionsResult Object
                               * https://developers.google.com/maps/documentation/javascript/directions?hl=en#DirectionsResults
                               */
                    // Custom Markers
                    if (_this.markerOptions !== undefined) {
                        var /** @type {?} */ _route = response.routes[0].legs[0];
                        // Origin Marker
                        // Origin Marker
                        _this.markerOptions.origin.map = map;
                        _this.markerOptions.origin.position = _route.start_location;
                        _this.originMarker = _this.setMarker(map, _this.originMarker, _this.markerOptions.origin, _route.start_address);
                        // Destination Marker
                        // Destination Marker
                        _this.markerOptions.destination.map = map;
                        _this.markerOptions.destination.position = _route.end_location;
                        _this.destinationMarker = _this.setMarker(map, _this.destinationMarker, _this.markerOptions.destination, _route.end_address);
                    }
                    _this.onChange.emit(response);
                }
            });
        });
    };
    /**
     * Custom Origin and Destination Icon
     *
     * \@memberof AgmDirection
     * @param {?} map
     * @param {?} marker
     * @param {?} markerOpts
     * @param {?} content
     * @return {?}
     */
    AgmDirection.prototype.setMarker = /**
     * Custom Origin and Destination Icon
     *
     * \@memberof AgmDirection
     * @param {?} map
     * @param {?} marker
     * @param {?} markerOpts
     * @param {?} content
     * @return {?}
     */
    function (map, marker, markerOpts, content) {
        var /** @type {?} */ infowindow = new google.maps.InfoWindow({
            content: content,
        });
        marker = new google.maps.Marker(markerOpts);
        marker.addListener('click', function () {
            infowindow.open(map, marker);
        });
        return marker;
    };
    AgmDirection.decorators = [
        { type: Directive, args: [{
                    selector: 'agm-direction'
                },] },
    ];
    /** @nocollapse */
    AgmDirection.ctorParameters = function () { return [
        { type: GoogleMapsAPIWrapper, },
    ]; };
    AgmDirection.propDecorators = {
        "origin": [{ type: Input },],
        "destination": [{ type: Input },],
        "travelMode": [{ type: Input },],
        "transitOptions": [{ type: Input },],
        "drivingOptions": [{ type: Input },],
        "waypoints": [{ type: Input },],
        "optimizeWaypoints": [{ type: Input },],
        "provideRouteAlternatives": [{ type: Input },],
        "avoidHighways": [{ type: Input },],
        "avoidTolls": [{ type: Input },],
        "renderOptions": [{ type: Input },],
        "visible": [{ type: Input },],
        "panel": [{ type: Input },],
        "markerOptions": [{ type: Input },],
        "onChange": [{ type: Output },],
    };
    return AgmDirection;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var AgmDirectionModule = (function () {
    function AgmDirectionModule() {
    }
    /**
     * @return {?}
     */
    AgmDirectionModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: AgmDirectionModule,
        };
    };
    AgmDirectionModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [
                        AgmDirection,
                    ],
                    exports: [
                        AgmDirection,
                    ]
                },] },
    ];
    return AgmDirectionModule;
}());

export { AgmDirectionModule, AgmDirection };
